### ISSUE TITLE
Why usecase controller is extending an infrastructure class?

### ISSUE THREAD

**Issue by imsergiobernal** [stemmlerjs/ddd-forum#14]
App layer shouldn't be aware of infrastructure implementation, why we extend an infrastructure implementation class?

I thought that the usecase controller shouldn't know about if the usecase is implemented through websocket, HTTP, TCP/IP or Message Broker; this makes the BaseController coupled to the HTTP protocol (VERBS).

I think the idea would be to return always a Result<ControllerResponse>, **develop a infrastructure implementation like HTTPRequestHandler** which requires a UseCase's Controller with HTTP verbs as methods, and then reuse it by Express, Koa... even at the same runtime or not. (library migration, performance testing...)

We could create another TCPRequestController, MQRequestController, which all of them requires the same as HTTPRequestHandler.

I tried to deal with this but I've found a concern:
What kind of response should the usecase controller return?

**Comment by stemmlerjs** [comment]
@imsergiobernal 

> App layer shouldn't be aware of infrastructure implementation, why we extend an infrastructure implementation class?

While that is correct, I'm not sure where the offending code you're referring to is.
Could you point me to where it is?

**Comment by stemmlerjs** [comment]
@imsergiobernal Your English is great! :)

So, here's my thoughts.

A `controller` is an _infrastructure_ layer concern, not an _application_ layer concern. And the `BaseController` class is also part of infrastructure. 

Web servers, caches, databases - anything to do with the outside world, are infra.

In our codebase, every controller _is a_ `BaseController`, which is just a nice base class encapsulating HTTP response logic with [intention revealing](https://khalilstemmler.com/articles/typescript-domain-driven-design/intention-revealing-interfaces/) methods.

You mentioned a `usecase controller`. Maybe you and I see it differently, but a `Use Case` is an application layer concern, and a `controller` is an infrastructure one. 

Therefore, a `use case controller` doesn't exist and is contradictory to the definitions of both.

We do, however, inject via [Dependency Injection](https://khalilstemmler.com/articles/tutorials/dependency-injection-inversion-explained/), use cases (like `UpvoteComment`) _into_ controllers (like `UpvoteCommentController`), which is **valid** to the [Dependency Rule](https://khalilstemmler.com/wiki/dependency-rule/).

**Comment by imsergiobernal** [comment]
So,

why the Controller is inside Usecase's folder and not inside infra folder?
Could you re-use the controller with another framework?
Could you re-use the controller with another request interface (Infrastructure Event for instance)?

👍

**Comment by stemmlerjs** [comment]
> why the Controller is inside Usecase's folder and not inside infra folder?

There's an architectural principle called **Package By Component**. 

![xh9i](https://user-images.githubusercontent.com/6892666/67135521-9b329180-f1e8-11e9-8bcd-462a6c2627e0.png)

Notice how easy it is to determine exactly what the features of this app are? That's called **package by component/module**.  When we do this, the app is practically telling us what it can do, without needing to look deep inside the files and classes.

Another term for this is screaming architecture. The app is screaming at us with the actors, use cases and domain that it addresses.

The alternative to this is **Package By Infrastructure*. 

> why the Controller is inside Usecase's folder and not inside ~~infra~~ application folder

Package by infrastructure is when you organize code according to the type of construct it is (`controllers`, `application layer`, `infra`, etc). When we do that, it hurts our ability to quickly identify where features are, and features get split across folders and files.

I really punched down on Uncle Bob's approach to designing applications by centering them around the use cases. That means that every time I start a new project, I'll identify the `actors`, the `use cases` and which `subdomain` those actors and use cases belong to.

When we package by component we get +1 readability, +1 cohesion (cohesive units of code).

Let's look a little bit deeper at the createPost/ folder for example.

![xh9k](https://user-images.githubusercontent.com/6892666/67135574-27dd4f80-f1e9-11e9-99ce-5e4a4c14ea24.png)

Here's really what I mean. In this cohesive component, we have everything we need in order to support the CreatePost.ts use case.

We have a:
- controller (to hook it up to infrastructure, express.js web server)
- dto (to specify the inputs to the use case)
- errors namespace (to define all the ways that the use case can succeed or fail)

And let's say for some reason we wanted to hook this use case up to a serverless function.

Depending on the service, we might be able to just export the controller and use that, but we might also need to define another adapter in order for the use case to be able to be used by the external infrastructure concern.

My point is: **everything is centered around the use cases of the application**.

It makes sense to co-locate things that are used together and their entire purpose of being is to work with each other (controller, dto, error namespace, use case).

This is the very essence of the Single Responsibility Principle.

> Could you re-use the controller with another framework? Could you re-use the controller with another request interface (Infrastructure Event for instance)?

I don't think you'd want to do that. What's being exported from the module is the **express controller** and the **use case** itself.

```typescript
import { DownvoteComment } from "./DownvoteComment";
import { postRepo, memberRepo, commentRepo, commentVotesRepo } from "../../../repos";
import { postService } from "../../../domain/services";
import { DownvoteCommentController } from "./DownvoteCommentController";

const downvoteComment = new DownvoteComment(
  postRepo, memberRepo, commentRepo, commentVotesRepo, postService
);

const downvoteCommentController = new DownvoteCommentController(
  downvoteComment
)

// we export the express controller and the use case itself
export {
  downvoteComment,                  
  downvoteCommentController 
}
```

The **use case** is what has reusability because it's an application layer concern. The express controller does not. It has a specific purpose: to be hooked up to an express RESTful API.

If you for some reason needed to support another delivery mechanism, like GraphQL or SOAP, you could dependency inject the `DownvoteComment` use case into a GraphQLController or a SoapController (each other specific **infrastructure layer concerns**).

**Comment by imsergiobernal** [comment]
@stemmlerjs 

Clean Architecture book, Use Case concept, page 188

> Notice also that the use case does not describe the user interface other than to
informally specify the data coming in from that interface, and the data going back
out through that interface. From the use case, it is impossible to tell whether the
application is delivered on the web, or on a thick client, or on a console, or is a pure
service.
This is very important. Use cases do not describe how the system appears to the user.
Instead, they describe the application-specific rules that govern the interaction
between the users and the Entities. How the data gets in and out of the system is
irrelevant to the use cases.

I throw the following question. Can I reuse this use-case to a Console output without duplicating?

**Comment by stemmlerjs** [comment]
@imsergiobernal Absolutely. To reuse it, you can import the `DownvoteComment` instance as shown in my example above and then call `downvoteComment.execute()`. 

No need to re-write the entire use case.
